integration/build/build_test.go

```diff
@@ -197,3 +197,71 @@ func TestBuildWithEmptyLayers(t *testing.T) {
 	resp.Body.Close()
 	require.NoError(t, err)
 }
+
+// #35652
+func TestBuildOnBuildMultiStage(t *testing.T) {
+	defer setupTest(t)()
+	// test both metadata and layer based commands as they may be implemented differently
+	dockerfile := `FROM busybox AS stage1
+ONBUILD RUN echo 'foo' >somefile
+ONBUILD ENV bar=baz
+
+FROM stage1
+RUN cat somefile
+
+FROM stage1
+RUN cat somefile`
+
+	ctx := context.Background()
+	source := fakecontext.New(t, "",
+		fakecontext.WithDockerfile(dockerfile))
+	defer source.Close()
+
+	apiclient := testEnv.APIClient()
+	resp, err := apiclient.ImageBuild(ctx,
+		source.AsTarReader(t),
+		types.ImageBuildOptions{
+			Remove:      true,
+			ForceRemove: true,
+		})
+
+	out := bytes.NewBuffer(nil)
+	require.NoError(t, err)
+	_, err = io.Copy(out, resp.Body)
+	resp.Body.Close()
+	require.NoError(t, err)
+
+	assert.Contains(t, out.String(), "Successfully built")
```

Comment:
nit: perhaps we should check that the output shows `foo` to make it a bit more apparent that we verified the `ONBUILD echo 'foo' >somefile` was succesful

Reply:
The `cat somefile` command would fail if the `ONBUILD RUN` doesn't complete correctly.

Reply:
I understand; it's a bit of a "nit", but we create a file with a specific content, and test for a "side effect" to check if that step is executed (the side-effect being "the build doesn't fail"), but we don't check if the step actually executed correctly. Things I was thinking of that can go wrong;

- a bug is introduced in the builder, and a the `RUN` step is cached from the previous build stage
- a bug is introduced in the builder, and a the `RUN` step is not executed
- someone decides to "optimise" the test and remove the `RUN` step (as it's not apparent we need it)

Reply:
1) Caching the run from the previous stage is the correct behavior. It is cached already. There are other tests that check commands resetting between steps if you mean that, this test doesn't check or use that.
2) Then it would cause `cat` to fail. Testing a single `RUN` command is not the purpose of this test.
3) This is a regression test for a specific bug(linked in a comment) in the previous version and this is the reproducer. Most of the builder tests are constructed so that missing feature would cause the build itself to fail. If someone optimizes out the functionality of the test then they can just as well remove the other assertions.

Reply:
> Then it would cause cat to fail. Testing a single `RUN` command is not the purpose of this test.

No, such a bug wouldn't make the test fail, it would just skip the `RUN` and as a result, skip the verification that the `ONBUILD RUN` was executed;

```bash
docker  build -<<EOF
FROM busybox AS stage1
ONBUILD RUN echo 'foo' >somefile
ONBUILD ENV bar=baz

FROM stage1

FROM stage1
EOF

Sending build context to Docker daemon  2.048kB
Step 1/5 : FROM busybox AS stage1
 ---> 6ad733544a63
Step 2/5 : ONBUILD RUN echo 'foo' >somefile
 ---> Running in ec0180b2264d
Removing intermediate container ec0180b2264d
 ---> 02cbf6891873
Step 3/5 : ONBUILD ENV bar=baz
 ---> Running in 0f49da0de672
Removing intermediate container 0f49da0de672
 ---> 946cd74cc000
Step 4/5 : FROM stage1
# Executing 2 build triggers
 ---> Running in a7158ed3ddb3
Removing intermediate container a7158ed3ddb3
 ---> Running in a3227f5b3a3e
Removing intermediate container a3227f5b3a3e
 ---> 95fc7839814b
Step 5/5 : FROM stage1
# Executing 2 build triggers
 ---> Using cache
 ---> Using cache
 ---> 95fc7839814b
Successfully built 95fc7839814b
```

> This is a regression test for a specific bug(linked in a comment)

The comment doesn't give me a hint what's being tested, nor does git (hence my other comment https://github.com/moby/moby/pull/35672/files#r157744702);

```bash
$ git show --summary 
commit ba0e5c801038d25873b6bee1f2ed2137e4470fa1 (HEAD -> tonistiigi-onbuild-test, tonistiigi/onbuild-test)
Author: Tonis Tiigi <tonistiigi@gmail.com>
Date:   Fri Dec 1 10:31:03 2017 -0800

    Add testcase for onbuild command in multi stage build
    
    Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>
```

> If someone optimizes out the functionality of the test then they can just as well remove the other assertions.

For that reason, it's important to describe what we're testing for, so that _if_ someone arrives at the test, it should be clear what's being tested, and what the purpose is of every step in the test.

Just from the test _itself_, I would deduct;

- We want the build to be succesful (`assert.Contains(t, out.String(), "Successfully built")`)
- We want 3 images to be built (`assert.Equal(t, 3, len(imageIDs))`)
- We want to be sure the image's metadata (`ENV`) is set in the final image (`assert.Contains(t, image.Config.Env, "bar=baz")`)

Without understanding of what we're testing for, optimizing the test by removing the `RUN` and `ONBUILD` run would still make the test pass, and make it faster.

Theoretical? Unfortunately not; a good example is the [`TestEventsLimit()`](https://github.com/moby/moby/blob/402540708c9a0c35dc0b279a0f330455633537b8/integration-cli/docker_cli_events_test.go#L84-L126) test, which has been [slow, and flaky](https://github.com/moby/moby/issues/28873) [numerous times](https://github.com/moby/moby/search?q=TestEventsLimit&type=Issues&utf8=✓). Originally [using 30 containers](https://github.com/moby/moby/blob/6b5b3250e8f63d2fcbbe132a02b7e4df9d1b60a5/integration-cli/docker_cli_events_test.go#L34-L46) and [_over a minute_ to run](https://github.com/moby/moby/issues/12675). Fixes have been applied to [cleanup those containers](https://github.com/moby/moby/commit/da3d3b97ebe1e6fe1254367521c725ca12a5e61d), [take clock skew into account](https://github.com/moby/moby/commit/e424c54d9c4030dc3290701e8e76add372e09e08), improve performance by [parallelizing the test](https://github.com/moby/moby/pull/12719/commits/40779b28bb04dca6178ba17eaeb93265e2974ee1), [_reducing_](https://github.com/moby/moby/pull/26968/commits/728a26519051561da7d0310911831955fa42b1d5) [parallelization](https://github.com/moby/moby/commit/2ec808ec0822333f0807ee40916e3ba0e3b064ff) for platform slowness/bugs, [adjust the test for new limits](https://github.com/moby/moby/commit/59d45c384a2de7bca73296ce1471646db14cb0c8), and [adjust for the number of events per container having changed](https://github.com/moby/moby/commit/ddae20c032058a0fd42c34c2e9750ee8f6296ac8#diff-9e07f7c1f4d16eb428d628f85ee1a6f1R89) (which actually completely broke the test). Perhaps with a proper description, all these refactors and mistakes would not have been made, and someone [would've seen what was actually tested](https://github.com/moby/moby/blob/560eb07009d65fbcfb49b2fa8b04a2e3f4d647e8/server/server.go#L2464-L2474), and noticed that's already [covered by a unit test](https://github.com/moby/moby/blob/8d056423f8c433927089bd7eb6bc97abbc1ed502/events/events_test.go#L97-L99)

Reply:
> No, such a bug wouldn't make the test fail, it would just skip the RUN and as a result, skip the verification that the ONBUILD RUN was executed;

As mentioned above, checking `RUN` is not the purpose of this test. There are already tests that cover `RUN` executing in multi-stage builds as well. This is a common pattern, in `cli_build_test.go` alone, there are 447 checks like this. At least 95% use it exactly like this test - by constructing a command that would fail if the tested condition isn't true.

> The comment doesn't give me a hint what's being tested, 

It is in https://github.com/moby/moby/pull/35672/commits/ba0e5c801038d25873b6bee1f2ed2137e4470fa1#diff-1cf8c643831a82f47962fea059992ab2R201 . That issue gives the best information about what is tested. The test description could even be: "This sequence of commands didn't work in 17.09 so it was evident that we don't have coverage for it". I copied something from that issue to here but we can't expect anyone to refactor a regression test without opening up the issue it is linked against.

> Just from the test itself, I would deduct;

By following this logic one could remove most of `cli_build_test.go`. A lot of these tests have zero assertions.

> We want 3 images to be built (assert.Equal(t, 3, len(imageIDs)))

This test builds 2 images. That it builds 2 images from 3 stages is again covered by another test. This construction is needed because the old helpers for writing these tests were banned. I agree that the old tests were more readable but this is out of my control. I check the intermediate result here to avoid panicking if someone breaks this test and instead fail with assertion error that will help with debugging.

> Theoretical? Unfortunately not; a good example is the TestEventsLimit() test

`TestEventsLimit` doesn't have any link to the issue or problem it is fixing, neither does it have comments clarifying the possibly unclear parts of the test. So not sure how this example applies here. Not to mention that `TestEventsLimit` is a useless slow test but this is a fast test checking for the bug we shipped in a stable release.

Reply:
Mentioning an issue number really isn't a replacement for a commit message or comment in the code; code moves around, sometimes even between repositories, and the information gets lost.